### PR TITLE
Serial reader

### PR DESCRIPTION
**Pull request by k323r**

### CODE REVIEW

File: msb/serial/SerialReader.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    def read_message(self):
+        with serial.Serial(
+            port=self.config.device,
+            baudrate=self.config.baudrate,
+            timeout=self.config.timeout,
+            parity=serial.PARITY_NONE,
+            stopbits=serial.STOPBITS_ONE,
+            bytesize=serial.EIGHTBITS,
+        ) as serial_reader:
+            # Set RTS to low and DTR to high
```
Comment (kawindme) [inline msb/serial/SerialReader.py:54]: If this is specific to Vestas controllers, consider defining this in the config.

File: msb/serial/SerialReader.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+from msb.serial.config import SerialConf
+from msb.config import load_config
+from msb.zmq_base import get_default_publisher
+
+
+class SerialReader:
+    def __init__(self, config, publisher):
+        self.config = config
+        self.publisher = publisher
+        self.measurement_keys = ["pow", "gen", "rtr", "wnd", "pit"]
```
Comment (kawindme) [inline msb/serial/SerialReader.py:15]: This should probably be defined in the config.

File: msb/serial/SerialReader.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    def start_loop(self):
+        for message in self.read_message():
+            self.read_a_message_and_publish(message)
+
+    def extractFloats(self, text) -> str:
+        # Match regex, return a tuple
+        try:
+            matching_tuple = re.findall(self.pattern, text)
+            return matching_tuple[0]
+        except Exception as e:
```
Comment (kawindme) [inline msb/serial/SerialReader.py:41]: Explicitly state exception here. Also, `as e` is not necessary.

File: msb/serial/SerialReader.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            # Required for a read-only connection to Vestas controller
+            serial_reader.rts = False
+            serial_reader.dtr = True
+
+            buffer = ""
+
+            while True:
+                if serial_reader.in_waiting > 0:
+                    try:
+                        buffer = buffer + serial_reader.read(serial_reader.in_waiting).decode()
```
Comment (kawindme) [inline msb/serial/SerialReader.py:64]: Explicitly state encoding here., and add encoding to config.

File: msb/serial/SerialReader.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                        if "1:OVERVIEW" in buffer:
```
Comment (kawindme) [inline msb/serial/SerialReader.py:65]: Is this Vestas controller specific, as well?

File: msb/serial/SerialReader.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                            message = buffer
+                            buffer = ""
+                            yield message
+                    except UnicodeError as e:
```
Comment (kawindme) [inline msb/serial/SerialReader.py:69]: `as e` not used.

File: msb/serial/SerialReader.py (hunk 4 above)
Comment (kawindme) [inline msb/serial/SerialReader.py:64]: Consider using a list as buffer instead. Strings are immutable in python, so for every iteration the old string is discarded, and a new one is created. Using a list and concatenating all string only once should be faster.

File: msb/serial/SerialReader.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    def read_a_message_and_publish(self, message):
+        data_values = self.extractFloats(message)
+        payload = self.create_payload(data_values)
+        self.publisher.send(self.topic, payload)
+
+    def start_loop(self):
+        for message in self.read_message():
+            self.read_a_message_and_publish(message)
+
+    def extractFloats(self, text) -> str:
```
Comment (kawindme) [inline msb/serial/SerialReader.py:36]: snake_case

File: msb/serial/SerialReader.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        # Match regex, return a tuple
+        try:
+            matching_tuple = re.findall(self.pattern, text)
+            return matching_tuple[0]
+        except Exception as e:
+            print(f"Could not find match on {text}")
+            return ""
+
+    def read_message(self):
```
Comment (kawindme) [inline msb/serial/SerialReader.py:45]: Consider renaming to `receive_messages` (or `read_messages`). (Plural as it is a generator)

File: msb/serial/SerialReader.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+import serial
+from serial.serialutil import SerialException
+import sys
+
+from msb.serial.config import SerialConf
+from msb.config import load_config
+from msb.zmq_base import get_default_publisher
+
+
+class SerialReader:
```
Comment (kawindme) [inline msb/serial/SerialReader.py:11]: Consider reordering the methods to make it easier to follow. I.e. ascending or descending in call order.

File: msb/serial/SerialReader.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+        # make sure that topic is a byte string
+        if isinstance(self.topic, str):
+            self.topic = self.topic.encode()
+
+    def create_payload(self, data_values):
+        payload = {k: v for k, v in zip(self.measurement_keys, data_values)}
+        return payload
+
+    def read_a_message_and_publish(self, message):
```
Comment (kawindme) [inline msb/serial/SerialReader.py:27]: `read_message_and_publish`?

File: msb/serial/SerialReader.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+        data_values = self.extractFloats(message)
+        payload = self.create_payload(data_values)
+        self.publisher.send(self.topic, payload)
+
+    def start_loop(self):
```
Comment (kawindme) [inline msb/serial/SerialReader.py:32]: `run` or `start`?

File: msb/serial/config.py [hunk 12]
```diff
@@ -0,0 +1,10 @@
+from msb.config.MSBConfig import MSBConf
+
+
+class SerialConf(MSBConf):
+    regex: str = "(\\d*\\.?\\d+)kW\\s*(\\d*\\.?\\d+)rpm\\s*(\\d*\\.?\\d+)rpm\\s*(\\d*\\.?\\d+)m/s\\s*(-?\\d*\\.*\\d*)"
```
Comment (kawindme) [inline msb/serial/config.py:5]: Regex could probably be simplified using `r""` string. The question is how the regex has to be defined in the YAML config. 🤔

File: msb/serial/config.py [hunk 13]
```diff
@@ -0,0 +1,10 @@
...
+    device: str = "/usb/TTY0"
+    topic: str = "imu"
```
Comment (kawindme) [inline msb/serial/config.py:7]: The topic does not seem correct.

**Review by kawindme — CHANGES_REQUESTED** [review]

**Comment by pacjac** [comment]
I would advise against merging this branch, as it is dated and reflects design that is specific to an ongoing project. A more generalized serial reader will be implemented.

**Pull request closed without merging**